### PR TITLE
PostReaction struct improvements

### LINKED ISSUE #157: Improve the structure of PostReaction
## Context
Currently in order to represent a single post reaction, what we are doing is using this object: 

```go
// PostReaction is a struct of a user reaction to a post
type PostReaction struct {
  Owner sdk.AccAddress `json:"owner" yaml:"owner"` // Creator that has created the reaction
  Value string         `json:"value" yaml:"value"` // Value of the reaction, either an emoji or a shortcode
}
```

This is used also inside the `GenesisData` to represent all the reactions that have been added to the different posts. 

Using this object from an outsider point of view, it is not enough in specific cases. Particularly, I've found some difficulties when handling it inside the genesis from within [DJuno](https://github.com/desmos-labs/djuno). Since the `Value` field can contain either the reaction shortocode or emoji value, it's hard to parse them and store them inside the database. 

## Solution
In order to solve the above mentioned problem I propose a revisit of this data type, in order to extend it so that it includes both the shortcode and the value of the emoji: 

```go
// PostReaction is a struct of a user reaction to a post
type PostReaction struct {
  Owner sdk.AccAddress `json:"owner" yaml:"owner"` // Creator that has created the reaction
  Shortcode string     `json:"shortcode" yaml:"shortcode"` // Shortcode of the reaction
  Value string         `json:"value" yaml:"value"` // Value of the reaction, either an emoji or an URI
}
```

### Notes
I acknowledge that this approach comes with a con, which is the fact that emoji values and registered emojis will occupy a lot more storage than the current version due to the fact that their shortcode will be stored multiple times.

While this might be a problem, supposing we have all reactions with shortcodes that are 20 characters long (equivalent of 160 bytes, which is a very long shortcode to be honest), it would mean that in order to have a significant disk space change (1GB) we would have to have at least 6,250,000 reactions used. Personally I think that 1GB for 6 millions reactions is fair as long as that change allows for a better usage from the developer perspective.

### PR DESCRIPTION
**Pull request by bragaz**
## Description
This PR improves the `PostReaction` type structure to allow a better integration with middle layer applications.
Closes #157.

## Checklist
- [x] Targeted PR against correct branch.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote unit tests.
- [x] Wrote integration tests (simulation & CLI). 
- [x] Updated the documentation. 
- [x] Added an entry to the `CHANGELOG.md` file.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Version 0.6.0
 ## Changes
 - Implemented tags in post medias (#118)
-
+- Edited PostReaction struct to allow a better integration with middle layer applications (#157)
```
Comment (RiccardoM) [inline CHANGELOG.md:4]: Better adding a blank line after this line

```suggestion
- Edited PostReaction struct to allow a better integration with middle layer applications (#157)

```

File: x/posts/internal/keeper/querier.go [hunk 2]
```diff
@@ -45,16 +44,9 @@ func getPostResponse(ctx sdk.Context, keeper Keeper, post types.Post) types.Post
...
 	var reactionsResponses = make([]types.ReactionQueryResponse, len(postReactions))
 	for index, reaction := range postReactions {
-		var reactionValue string
-		if em, err := emoji.LookupEmojiByCode(reaction.Value); err == nil {
-			reactionValue = em.Value
-		} else {
-			registeredReaction, _ := keeper.GetRegisteredReaction(ctx, reaction.Value, post.Subspace)
-			reactionValue = registeredReaction.Value
-		}
 		reactionsResponses[index] = types.ReactionQueryResponse{
```
Comment (RiccardoM) [inline x/posts/internal/keeper/querier.go:47]: Since the new `PostReaction` type is pretty much equal to the `ReactionQueryResponse` type, I think that we can delete the latter and leave only the `PostReaction` type to be used inside the `PostQueryResponse` object.

File: x/posts/internal/types/models/post_response_test.go [hunk 3]
```diff
@@ -82,7 +82,7 @@ func TestPostQueryResponse_MarshalJSON(t *testing.T) {
...
+			expResponse: `{"id":"dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","parent_id":"","message":"Post","created":"2020-02-02T15:00:00Z","last_edited":"0001-01-01T00:00:00Z","allows_comments":true,"subspace":"4e188d9c17150037d5199bbdb91ae1eb2a78a15aca04cb35530cccb81494b36e","creator":"cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47","medias":[{"uri":"https://uri.com","mime_type":"text/plain","tags":["cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47"]}],"poll_data":{"question":"poll?","provided_answers":[{"id":"1","text":"Yes"},{"id":"2","text":"No"}],"end_date":"2050-01-01T15:15:00Z","is_open":true,"allows_multiple_answers":false,"allows_answer_edits":true},"poll_answers":[{"answers":["1"],"user":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4"}],"reactions":[{"owner":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4","shortcode":"https://example.com/like","value":":like:"},{"owner":"cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae","shortcode":"👍","value":":+1:"}],"children":["dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1"]}`,
```
Comment (RiccardoM) [inline x/posts/internal/types/models/post_response_test.go:85]: Please revert this as it should not change. The error is in lines 67-69

```suggestion
			expResponse: `{"id":"dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","parent_id":"","message":"Post","created":"2020-02-02T15:00:00Z","last_edited":"0001-01-01T00:00:00Z","allows_comments":true,"subspace":"4e188d9c17150037d5199bbdb91ae1eb2a78a15aca04cb35530cccb81494b36e","creator":"cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47","medias":[{"uri":"https://uri.com","mime_type":"text/plain","tags":["cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47"]}],"poll_data":{"question":"poll?","provided_answers":[{"id":"1","text":"Yes"},{"id":"2","text":"No"}],"end_date":"2050-01-01T15:15:00Z","is_open":true,"allows_multiple_answers":false,"allows_answer_edits":true},"poll_answers":[{"answers":["1"],"user":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4"}],"reactions":[{"value":"https://example.com/like","shortcode":":like:","owner":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4"},{"value":"👍","shortcode":":+1:","owner":"cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae"}],"children":["dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1"]}`,
```

File: x/posts/internal/types/models/post_response_test.go [hunk 4]
```diff
@@ -64,9 +64,9 @@ func TestPostQueryResponse_MarshalJSON(t *testing.T) {
 		"dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1",
 	}
 
-	reactionsResponses := []models.ReactionQueryResponse{
-		models.NewReactionQueryResponse("https://example.com/like", ":like:", liker),
-		models.NewReactionQueryResponse("👍", ":+1:", otherLiker),
+	reactionsResponses := []models.PostReaction{
+		models.NewPostReaction("https://example.com/like", ":like:", liker),
+		models.NewPostReaction("👍", ":+1:", otherLiker),
```
Comment (RiccardoM) [inline x/posts/internal/types/models/post_response_test.go:69]: These lines are wrong, the `NewPostReaction` first parameter is the shortcode, and then the value. Not the opposite: 

```suggestion
		models.NewPostReaction(":like:", "https://example.com/like", liker),
		models.NewPostReaction(":+1:", "👍", otherLiker),
```

File: x/posts/internal/types/models/post_response_test.go [hunk 5]
```diff
@@ -92,7 +92,7 @@ func TestPostQueryResponse_MarshalJSON(t *testing.T) {
...
+			expResponse: `{"id":"dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","parent_id":"","message":"Post","created":"2020-02-02T15:00:00Z","last_edited":"0001-01-01T00:00:00Z","allows_comments":true,"subspace":"4e188d9c17150037d5199bbdb91ae1eb2a78a15aca04cb35530cccb81494b36e","creator":"cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47","medias":[{"uri":"https://uri.com","mime_type":"text/plain"}],"poll_answers":[{"answers":["1"],"user":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4"}],"reactions":[{"owner":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4","shortcode":"https://example.com/like","value":":like:"},{"owner":"cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae","shortcode":"👍","value":":+1:"}],"children":["dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1"]}`,
```
Comment (RiccardoM) [inline x/posts/internal/types/models/post_response_test.go:95]: Please revert this as it shouldn't change

```suggestion
			expResponse: `{"id":"dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","parent_id":"","message":"Post","created":"2020-02-02T15:00:00Z","last_edited":"0001-01-01T00:00:00Z","allows_comments":true,"subspace":"4e188d9c17150037d5199bbdb91ae1eb2a78a15aca04cb35530cccb81494b36e","creator":"cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47","medias":[{"uri":"https://uri.com","mime_type":"text/plain"}],"poll_answers":[{"answers":["1"],"user":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4"}],"reactions":[{"value":"https://example.com/like","shortcode":":like:","owner":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4"},{"value":"👍","shortcode":":+1:","owner":"cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae"}],"children":["dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1"]}`,
```

File: x/posts/internal/types/models/post_response_test.go [hunk 6]
```diff
@@ -102,7 +102,7 @@ func TestPostQueryResponse_MarshalJSON(t *testing.T) {
...
+			expResponse: `{"id":"dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","parent_id":"","message":"Post","created":"2020-02-02T15:00:00Z","last_edited":"0001-01-01T00:00:00Z","allows_comments":true,"subspace":"4e188d9c17150037d5199bbdb91ae1eb2a78a15aca04cb35530cccb81494b36e","creator":"cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47","medias":[{"uri":"https://uri.com","mime_type":"text/plain","tags":["cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47"]}],"reactions":[{"owner":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4","shortcode":"https://example.com/like","value":":like:"},{"owner":"cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae","shortcode":"👍","value":":+1:"}],"children":["dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1"]}`,
```
Comment (RiccardoM) [inline x/posts/internal/types/models/post_response_test.go:105]: Please revert this as it shouldn't change

```suggestion
			expResponse: `{"id":"dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","parent_id":"","message":"Post","created":"2020-02-02T15:00:00Z","last_edited":"0001-01-01T00:00:00Z","allows_comments":true,"subspace":"4e188d9c17150037d5199bbdb91ae1eb2a78a15aca04cb35530cccb81494b36e","creator":"cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47","medias":[{"uri":"https://uri.com","mime_type":"text/plain","tags":["cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47"]}],"reactions":[{"value":"https://example.com/like","shortcode":":like:","owner":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4"},{"value":"👍","shortcode":":+1:","owner":"cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae"}],"children":["dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1"]}`,
```

File: x/posts/internal/types/models/post_response_test.go [hunk 7]
```diff
@@ -112,7 +112,7 @@ func TestPostQueryResponse_MarshalJSON(t *testing.T) {
...
+			expResponse: `{"id":"dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","parent_id":"","message":"Post","created":"2020-02-02T15:00:00Z","last_edited":"0001-01-01T00:00:00Z","allows_comments":true,"subspace":"4e188d9c17150037d5199bbdb91ae1eb2a78a15aca04cb35530cccb81494b36e","creator":"cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47","poll_data":{"question":"poll?","provided_answers":[{"id":"1","text":"Yes"},{"id":"2","text":"No"}],"end_date":"2050-01-01T15:15:00Z","is_open":true,"allows_multiple_answers":false,"allows_answer_edits":true},"poll_answers":[{"answers":["1"],"user":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4"}],"reactions":[{"owner":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4","shortcode":"https://example.com/like","value":":like:"},{"owner":"cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae","shortcode":"👍","value":":+1:"}],"children":["dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1"]}`,
```
Comment (RiccardoM) [inline x/posts/internal/types/models/post_response_test.go:115]: Please revert this as it shouldn't change

```suggestion
			expResponse: `{"id":"dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","parent_id":"","message":"Post","created":"2020-02-02T15:00:00Z","last_edited":"0001-01-01T00:00:00Z","allows_comments":true,"subspace":"4e188d9c17150037d5199bbdb91ae1eb2a78a15aca04cb35530cccb81494b36e","creator":"cosmos1y54exmx84cqtasvjnskf9f63djuuj68p7hqf47","poll_data":{"question":"poll?","provided_answers":[{"id":"1","text":"Yes"},{"id":"2","text":"No"}],"end_date":"2050-01-01T15:15:00Z","is_open":true,"allows_multiple_answers":false,"allows_answer_edits":true},"poll_answers":[{"answers":["1"],"user":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4"}],"reactions":[{"value":"https://example.com/like","shortcode":":like:","owner":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4"},{"value":"👍","shortcode":":+1:","owner":"cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae"}],"children":["dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1","dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1"]}`,
```

File: x/posts/internal/types/models/post_response_test.go [hunk 8]
```diff
@@ -167,21 +167,19 @@ func TestPostQueryResponse_String(t *testing.T) {
 		models.NewUserAnswers(
 			models.NewUserAnswer([]models.AnswerID{models.AnswerID(1)}, liker),
 		),
-		[]models.ReactionQueryResponse{
-			models.NewReactionQueryResponse("https://example.com/like", ":like:", liker),
-			models.NewReactionQueryResponse("👍", ":+1:", otherLiker),
+		[]models.PostReaction{
+			models.NewPostReaction("https://example.com/like", ":like:", liker),
+			models.NewPostReaction("👍", ":+1:", otherLiker),
```
Comment (RiccardoM) [inline x/posts/internal/types/models/post_response_test.go:172]: The `NewPostReaction` method is wrongly used here. The first parameter is the shortcode, and the second one is the value, not the other way around. Please fix it

```suggestion
			models.NewPostReaction(":like:", "https://example.com/like", liker),
			models.NewPostReaction(":+1:", "👍", otherLiker),
```

File: x/posts/internal/types/models/post_response_test.go [hunk 9]
```diff
@@ -167,21 +167,19 @@ func TestPostQueryResponse_String(t *testing.T) {
...
 	)
 
-	expected := `
-ID: dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1
-Reactions: [{https://example.com/like :like: cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4} {👍 :+1: cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae}]
-Children: [dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1, dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1]
-`
+	expected := `ID: dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1
+Reactions: [{"owner":"cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4","shortcode":"https://example.com/like","value":":like:"} {"owner":"cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae","shortcode":"👍","value":":+1:"}]
+Children: [dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1, dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1]`
```
Comment (RiccardoM) [inline x/posts/internal/types/models/post_response_test.go:182]: Please revert these lines as they should not change

Comment (bragaz) [reply x/posts/internal/types/models/post_response_test.go:182]: everything should be fine now, there was a problem inside the `PostReaction` type where its fields are alfabetically ordered (probably due to a refactor that I've done).

Comment (RiccardoM) [reply x/posts/internal/types/models/post_response_test.go:182]: There is still something fishy about the tests. You are creating the reactions as follows: 

```golang
[]models.PostReaction{
  models.NewPostReaction(":like:", "https://example.com/like", liker),
  models.NewPostReaction(":+1:", "👍", otherLiker),
},
```

But then you are testing if the result is 

```json
[
  {
    "owner": "cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4",
    "shortcode": "https://example.com/like",
    "value": ":like:"
  }
  {
    "owner": "cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae",
    "shortcode": "👍",
    "value": ":+1:"
  }
]
```

As you can see in the result the `value` and `shortcode` values are inverted. There is either an error in the test or an error somewhere else. Please make the test expected result correct (invert those values) and fix the bug accordingly.

Comment (bragaz) [reply x/posts/internal/types/models/post_response_test.go:182]: yeah I've still need to commit the changes
